routing/routing_helpers.cpp

```diff
@@ -57,14 +57,24 @@ void ReconstructRoute(IDirectionsEngine & engine, RoadGraphBase const & graph,
   route.SetAltitudes(move(altitudes));
 
   vector<traffic::SpeedGroup> traffic;
-  if (trafficColoring && !trafficSegs.empty())
+  if (trafficStash && !trafficSegs.empty())
   {
     traffic.reserve(2 * trafficSegs.size());
-    for (TrafficInfo::RoadSegmentId const & seg : trafficSegs)
+    for (Segment const & seg : trafficSegs)
     {
-      auto const it = trafficColoring->find(seg);
-      SpeedGroup const segTraffic = (it == trafficColoring->cend()) ? SpeedGroup::Unknown
-                                                               : it->second;
+      traffic::TrafficInfo::RoadSegmentId roadSegment(
+          seg.GetFeatureId(), seg.GetSegmentIdx(),
+          seg.IsForward() ? traffic::TrafficInfo::RoadSegmentId::kForwardDirection
+                          : traffic::TrafficInfo::RoadSegmentId::kReverseDirection);
+
+      auto segTraffic = SpeedGroup::Unknown;
+      auto trafficColoring = trafficStash->Get(seg.GetMwmId());
```

Comment:
Am I right that in this loop over trafficSegs there will be a long consecutive runs with the same mwm id? I mean, on the route, mwm is changed quite rarely. I suggest you to use this fact and rewrite the loop. For example:
```c++
size_t i = 0;
while (i != trafficSegs.size())
{
  size_t j = i;
  while (j != trafficSegs.size() && trafficSegs[i].GetMwmId() == trafficSegs[j].GetMwmId())
    ++j;
  // All segs from the range [i, j) have the same mwm id.
}